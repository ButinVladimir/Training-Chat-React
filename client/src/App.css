.chat-app {
  display: flex;
  width: 100vw;
  max-width: 100vw;
  height: 100vh;
  max-height: 100vh;
  flex-direction: row;
  flex-wrap: nowrap;
}

.chat-app-column {
  display: flex;
  flex-direction: column;
  flex-wrap: nowrap;
  flex-grow: 1;
  max-height: 100vh;
}

.chat-app-row {
  border: 1px solid black;
  flex-grow: 1;
  position: relative;
  display: flex;
  align-items: stretch;
  justify-content: stretch;
}
