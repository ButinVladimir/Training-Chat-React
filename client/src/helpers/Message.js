import PropTypes from 'prop-types';

class Message {
  constructor(id, date, from, to, message) {
    this.id = id;
    this.date = date;
    this.from = from;
    this.to = to;
    this.message = message;
  }
}

const messageShape = {
  id: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
  from: PropTypes.string.isRequired,
  to: PropTypes.arrayOf(PropTypes.string),
  message: PropTypes.string.isRequired,
};

export default Message;
export { messageShape };
