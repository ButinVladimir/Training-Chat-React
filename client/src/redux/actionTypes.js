export const CHANGE_LOGIN = 'CHANGE_LOGIN';
export const LOGIN = 'LOGIN';
export const LOGOUT = 'LOGOUT';
export const CHANGE_SPEECH = 'CHANGE_SPEECH';
export const SAY = 'SAY';
export const ADD_TO = 'ADD_TO';
export const REMOVE_TO = 'REMOVE_TO';
export const CONNECT = 'CONNECT';
export const DISCONNECT = 'DISCONNECT';
export const SHOW_ERROR = 'SHOW_ERROR';
export const HIDE_ERROR = 'HIDE_ERROR';
export const USER_APPEARS = 'USER_APPEARS';
export const USER_DISAPPEARS = 'USER_DISAPPEARS';
export const GET_MESSAGE = 'GET_MESSAGE';
export const UPDATE_USER_LIST = 'UPDATE_USER_LIST';
