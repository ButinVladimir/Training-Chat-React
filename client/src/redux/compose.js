export default function compose(...args) {
  return (state, action) => {
    let resultState = state;
    args.forEach((reducer) => { resultState = reducer(resultState, action); });

    return resultState;
  };
}
