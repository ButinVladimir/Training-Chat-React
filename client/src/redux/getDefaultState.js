export default () => ({
  loginPanel: {
    loginned: false,
    login: '',
  },
  usersPanel: {
    usersList: [],
  },
  speechPanel: {
    speech: '',
    to: [],
  },
  chatPanel: {
    messages: [],
  },
  connected: false,
  errorPanel: {
    showError: false,
    error: '',
  },
});
