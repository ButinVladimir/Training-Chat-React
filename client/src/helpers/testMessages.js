import Message from './Message';

const testMessages = [
  new Message('1', '8/28/2018', 'User 1', [], 'Message 1'),
  new Message('2', '8/28/2018', 'User 1', [], 'Message 2'),
  new Message('3', '8/28/2018', 'User 1', ['User 2'], 'Message 3 to one user'),
  new Message('4', '8/28/2018', 'User 2', ['User 1', 'User 2'], 'User 2 answer'),
  new Message('5', '8/28/2018', 'User 3', ['User 1', 'User 2'], 'User 3 question?'),
  new Message('6', '8/28/2018', 'User 1', ['User 1', 'User 2', 'User 3'], 'User 1 answer'),
  new Message('7', '8/28/2018', 'User 2', [], 'Message 7'),
  new Message('8', '8/28/2018', 'User 2', [], 'Message 8'),
  new Message('9', '8/28/2018', 'User 1', [], 'Message 9'),
  new Message('10', '8/28/2018', 'User 3', [], 'Message 10'),
  new Message('11', '8/28/2018', 'User 1', [], 'Message 1'),
  new Message('12', '8/28/2018', 'User 1', [], 'Message 2'),
  new Message('13', '8/28/2018', 'User 1', ['User 2'], 'Message 3 to one user'),
  new Message('14', '8/28/2018', 'User 2', ['User 1', 'User 2'], 'User 2 answer'),
  new Message('15', '8/28/2018', 'User 3', ['User 1', 'User 2'], 'User 3 question?'),
  new Message('16', '8/28/2018', 'User 1', ['User 1', 'User 2', 'User 3'], 'User 1 answer'),
  new Message('17', '8/28/2018', 'User 2', [], 'Message 7'),
  new Message('18', '8/28/2018', 'User 2', [], 'Message 8'),
  new Message('19', '8/28/2018', 'User 1', [], 'Message 9'),
  new Message('20', '8/28/2018', 'User 3', [], 'Message 10'),
  new Message('21', '8/28/2018', 'User 1', [], 'Message 1'),
  new Message('22', '8/28/2018', 'User 1', [], 'Message 2'),
  new Message('23', '8/28/2018', 'User 1', ['User 2'], 'Message 3 to one user'),
  new Message('24', '8/28/2018', 'User 2', ['User 1', 'User 2'], 'User 2 answer'),
  new Message('25', '8/28/2018', 'User 3', ['User 1', 'User 2'], 'User 3 question?'),
  new Message('26', '8/28/2018', 'User 1', ['User 1', 'User 2', 'User 3'], 'User 1 answer'),
  new Message('27', '8/28/2018', 'User 2', [], 'Message 7'),
  new Message('28', '8/28/2018', 'User 2', [], 'Message 8'),
  new Message('29', '8/28/2018', 'User 1', [], 'Message 9'),
  new Message('30', '8/28/2018', 'User 3', [], 'Message 10'),
];

export default testMessages;
