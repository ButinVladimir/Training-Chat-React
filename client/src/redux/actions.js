import * as actionTypes from './actionTypes';

export const changeLogin = loginValue => ({
  type: actionTypes.CHANGE_LOGIN,
  loginValue,
});

export const login = () => ({
  type: actionTypes.LOGIN,
});

export const logout = () => ({
  type: actionTypes.LOGOUT,
});

export const changeSpeech = speechValue => ({
  type: actionTypes.CHANGE_SPEECH,
  speechValue,
});

export const say = (id, date) => ({
  type: actionTypes.SAY,
  id,
  date,
});

export const addTo = recipient => ({
  type: actionTypes.ADD_TO,
  recipient,
});

export const removeTo = recipient => ({
  type: actionTypes.REMOVE_TO,
  recipient,
});
