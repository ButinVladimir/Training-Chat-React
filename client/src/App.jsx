import React from 'react';
import ChatPanel from './components/chatPanel/ChatPanel';
import SpeechPanel from './components/speechPanel/SpeechPanelContainer';
import LoginPanel from './components/loginPanel/LoginPanelContainer';
import UsersPanel from './components/usersPanel/UsersPanelContainer';
import testMessages from './helpers/testMessages';
import './App.css';

const connected = true;

function App() {
  return (
    <div className="chat-app">
      <div className="chat-app-column grow">
        <div className="chat-app-row grow">
          <ChatPanel
            connected={connected}
          />
        </div>

        {connected && (
          <div className="chat-app-row">
            <SpeechPanel />
          </div>
        )}
      </div>

      {connected && (
        <div className="chat-app-column">
          <div className="chat-app-row">
            <LoginPanel />
          </div>
          <div className="chat-app-row grow">
            <UsersPanel />
          </div>
        </div>
      )}
    </div>);
}

export default App;
