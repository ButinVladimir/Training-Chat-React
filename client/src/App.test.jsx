import React from 'react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import App from './App';
import { connect, disconnect, showError } from './redux/actions';
import getDefaultState from './redux/getDefaultState';
import { USER_ERROR } from './helpers/socketEvents';

describe('App', () => {
  let mockStore;

  beforeEach(() => {
    mockStore = configureStore();
  });

  it('renders without crashing', () => {
    const store = mockStore(getDefaultState());
    const component = mount(
      <Provider store={store}>
        <App url='testUrl' />
      </Provider>
    );

    expect(component).toMatchSnapshot();
        
    const appComponent = component.find('App');    
    const instance = appComponent.instance();
    expect(appComponent.state('mounted')).toEqual(true);

    expect(instance.socket.__mockHandlers.has('connect')).toEqual(true);
    expect(instance.socket.__mockHandlers.has('reconnect')).toEqual(true);
    expect(instance.socket.__mockHandlers.has('disconnect')).toEqual(true);
    expect(instance.socket.__mockHandlers.has('connect_error')).toEqual(true);
    expect(instance.socket.__mockHandlers.has('reconnect_error')).toEqual(true);
    expect(instance.socket.__mockHandlers.has('error')).toEqual(true);
    expect(instance.socket.__mockHandlers.has(USER_ERROR)).toEqual(true);
  });

  it('connects', () => {
    const store = mockStore(getDefaultState());
    const component = mount(
      <Provider store={store}>
        <App url='testUrl' />
      </Provider>
    );

    let actions = store.getActions();
    let expectedActions = [];
    expect(actions).toEqual(expectedActions);

    const instance = component.find('App').instance();
    instance.socket.__mockHandlers.get('connect')();

    actions = store.getActions();
    expectedActions = [connect()];
    expect(actions).toEqual(expectedActions);
  });

  it('reconnects', () => {
    const store = mockStore(getDefaultState());
    const component = mount(
      <Provider store={store}>
        <App url='testUrl' />
      </Provider>
    );

    let actions = store.getActions();
    let expectedActions = [];
    expect(actions).toEqual(expectedActions);

    const instance = component.find('App').instance();
    instance.socket.__mockHandlers.get('reconnect')();

    actions = store.getActions();
    expectedActions = [disconnect()];
    expect(actions).toEqual(expectedActions);
  });

  it('disconnects', () => {
    const store = mockStore(getDefaultState());
    const component = mount(
      <Provider store={store}>
        <App url='testUrl' />
      </Provider>
    );

    let actions = store.getActions();
    let expectedActions = [];
    expect(actions).toEqual(expectedActions);

    const instance = component.find('App').instance();
    instance.socket.__mockHandlers.get('disconnect')();

    actions = store.getActions();
    expectedActions = [disconnect()];
    expect(actions).toEqual(expectedActions);
  });

  it('handles connect error', () => {
    const store = mockStore(getDefaultState());
    const component = mount(
      <Provider store={store}>
        <App url='testUrl' />
      </Provider>
    );

    let actions = store.getActions();
    let expectedActions = [];
    expect(actions).toEqual(expectedActions);

    const instance = component.find('App').instance();
    instance.socket.__mockHandlers.get('connect_error')();

    actions = store.getActions();
    expectedActions = [showError('Connect error')];
    expect(actions).toEqual(expectedActions);
  });

  it('handles reconnect error', () => {
    const store = mockStore(getDefaultState());
    const component = mount(
      <Provider store={store}>
        <App url='testUrl' />
      </Provider>
    );

    let actions = store.getActions();
    let expectedActions = [];
    expect(actions).toEqual(expectedActions);

    const instance = component.find('App').instance();
    instance.socket.__mockHandlers.get('reconnect_error')();

    actions = store.getActions();
    expectedActions = [showError('Reconnect error')];
    expect(actions).toEqual(expectedActions);
  });

  it('handles error', () => {
    const errorMessage = 'Some error occurred';
    const store = mockStore(getDefaultState());
    const component = mount(
      <Provider store={store}>
        <App url='testUrl' />
      </Provider>
    );

    let actions = store.getActions();
    let expectedActions = [];
    expect(actions).toEqual(expectedActions);

    const instance = component.find('App').instance();
    instance.socket.__mockHandlers.get('error')({message: errorMessage});

    actions = store.getActions();
    expectedActions = [showError(`Error: ${errorMessage}`)];
    expect(actions).toEqual(expectedActions);
  });

  it('handles user error', () => {
    const errorMessage = 'Some error occurred';
    const store = mockStore(getDefaultState());
    const component = mount(
      <Provider store={store}>
        <App url='testUrl' />
      </Provider>
    );

    let actions = store.getActions();
    let expectedActions = [];
    expect(actions).toEqual(expectedActions);

    const instance = component.find('App').instance();
    instance.socket.__mockHandlers.get(USER_ERROR)(errorMessage);

    actions = store.getActions();
    expectedActions = [showError(errorMessage)];
    expect(actions).toEqual(expectedActions);
  });

  it('handles unmounting', () => {
    const store = mockStore(getDefaultState());
    const component = mount(
      <Provider store={store}>
        <App url='testUrl' />
      </Provider>
    );

    const instance = component.find('App').instance();

    const disconnectMock = jest.fn();
    instance.socket.disconnect = disconnectMock;
    component.unmount();

    expect(disconnectMock).toHaveBeenCalledTimes(1);
    expect(disconnectMock).toHaveBeenCalledWith();
  });
});
