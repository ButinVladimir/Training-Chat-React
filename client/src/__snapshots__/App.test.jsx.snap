// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders without crashing 1`] = `
<div
  className="chat-app"
>
  <div
    className="chat-app-column grow"
  >
    <div
      className="chat-app-row grow"
    >
      <div
        className="chat-messages-list-container"
      >
        <ul
          className="chat-messages-list"
        />
      </div>
    </div>
    <div
      className="chat-app-row"
    >
      <div
        className="chat-speech-panel-container"
      >
        <div
          className="not-loginned-message"
        >
          Login first to send messages
        </div>
      </div>
    </div>
  </div>
  <div
    className="chat-app-column"
  >
    <div
      className="chat-app-row"
    >
      <div
        className="chat-login-panel-container"
      >
        <div
          className="form-row"
        >
          <input
            className="input"
            onChange={[Function]}
            type="text"
            value=""
          />
          <button
            className="login"
            onClick={[Function]}
            type="button"
          >
            Login
          </button>
        </div>
      </div>
    </div>
    <div
      className="chat-app-row grow"
    >
      <div
        className="chat-users-list-container"
      >
        <ul
          className="users-list"
        />
      </div>
    </div>
  </div>
</div>
`;
